source/extensions/filters/network/dubbo_proxy/active_message.cc

```diff
@@ -0,0 +1,407 @@
+}
+
+void ActiveMessageDecoderFilter::resetDownstreamConnection() {
+  parent_.resetDownstreamConnection();
+}
+
+void ActiveMessageDecoderFilter::resetStream() { parent_.resetStream(); }
+
+StreamInfo::StreamInfo& ActiveMessageDecoderFilter::streamInfo() { return parent_.streamInfo(); }
+
+// class ActiveMessage
+ActiveMessage::ActiveMessage(ConnectionManager& parent)
+    : parent_(parent), request_timer_(std::make_unique<Stats::Timespan>(
+                           parent_.stats().request_time_ms_, parent.time_system())),
+      request_id_(-1), stream_id_(parent.random_generator().random()),
+      stream_info_(parent.time_system()), pending_message_end_(false), local_response_sent_(false) {
+  parent_.stats().request_active_.inc();
+  stream_info_.setDownstreamLocalAddress(parent_.connection().localAddress());
+  stream_info_.setDownstreamRemoteAddress(parent_.connection().remoteAddress());
+}
+
+ActiveMessage::~ActiveMessage() {
+  parent_.stats().request_active_.dec();
+  request_timer_->complete();
+  for (auto& filter : decoder_filters_) {
+    filter->handler()->onDestroy();
+  }
+  ENVOY_LOG(debug, "ActiveMessage::~ActiveMessage()");
+}
+
+Network::FilterStatus ActiveMessage::transportBegin() {
+  filter_action_ = [](DubboFilters::DecoderFilter* filter) -> Network::FilterStatus {
+    return filter->transportBegin();
+  };
+
+  return this->applyDecoderFilters(nullptr);
+}
+
+Network::FilterStatus ActiveMessage::transportEnd() {
+  filter_action_ = [](DubboFilters::DecoderFilter* filter) -> Network::FilterStatus {
+    return filter->transportEnd();
+  };
+
+  return this->applyDecoderFilters(nullptr);
+}
+
+Network::FilterStatus ActiveMessage::messageBegin(MessageType type, int64_t message_id,
+                                                  SerializationType serialization_type) {
+  request_id_ = message_id;
+  filter_action_ = [type, message_id, serialization_type](
+                       DubboFilters::DecoderFilter* filter) -> Network::FilterStatus {
+    return filter->messageBegin(type, message_id, serialization_type);
+  };
+
+  return this->applyDecoderFilters(nullptr);
+}
+
+Network::FilterStatus ActiveMessage::messageEnd(MessageMetadataSharedPtr metadata) {
+  ASSERT(metadata->message_type() == MessageType::Request ||
+         metadata->message_type() == MessageType::Oneway);
+
+  // Currently only hessian serialization is implemented.
+  ASSERT(metadata->serialization_type() == SerializationType::Hessian);
+
+  ENVOY_LOG(debug, "dubbo request: start processing downstream request messages, id is {}",
+            metadata->request_id());
+
+  metadata_ = metadata;
+  filter_action_ = [metadata](DubboFilters::DecoderFilter* filter) -> Network::FilterStatus {
+    return filter->messageEnd(metadata);
+  };
+
+  auto status = applyDecoderFilters(nullptr);
+  if (status == Network::FilterStatus::StopIteration) {
+    pending_message_end_ = true;
+    return status;
+  }
+
+  finalizeRequest();
```

Comment:
Why not call this method in the last filter callback.

Reply:
1. since the Decoder has decoded the request data at this time, the statistical item in finalizeRequest function is specific to the request.
2. if StopIteration is returned in other filter callback, the decoder will stop decoding unless the continueDecoding interface is called.

Reply:
if a local reply is made in last filter callback, is this request still valid? If it is not valid, the finalizeRequest will perform stats statistics.

Reply:
If the last filter has a local reply, who is responsible for the deferred ActiveMessage?

Reply:
The finalizeRequest call has been moved to the last filter callback.